### PR TITLE
Use object based stuff. Add -d:vmathArrayBased switch to go back.

### PR DESCRIPTION
**Pull request by treeform**

### CODE REVIEW

File: src/vmath.nim [hunk 1]
```diff
@@ -132,137 +152,158 @@ genConstructor(uvec, UVec, uint32)
...
+
+  proc `x=`*[T](a: var GVec234[T], value: T) = a[0] = value
+  proc `y=`*[T](a: var GVec234[T], value: T) = a[1] = value
+  proc `z=`*[T](a: var GVec34[T], value: T) = a[2] = value
+  proc `w=`*[T](a: var GVec4[T], value: T) = a[3] = value
+
+else:
+
+  template `[]`*[T](a: GVec2[T], i: int): T =
+    cast[array[2, T]](a)[i]
```
Comment (guzba) [inline src/vmath.nim:192]: js?

File: src/vmath.nim [hunk 2]
```diff
@@ -271,36 +312,69 @@ genOp(`/`)
...
+      op(a[0], b)
+      op(a[1], b)
+      op(a[2], b)
+      op(a[3], b)
+
+else:
+  template genEqOp(op: untyped) =
+    proc op*[T](a: var GVec2[T], b: GVec2[T]) =
+      op(a.x, b.x)
+      op(a.y, b.y)
```
Comment (guzba) [inline src/vmath.nim:351]: still needed? vs above

File: src/vmath.nim [hunk 3]
```diff
@@ -434,60 +508,91 @@ type
...
+# proc `[]=`*[T](a: var GMat3[T], i, j: int, v: T) =
+#   cast[var array[3*3, T]](a)[i * 3 + j] = v
+# proc `[]=`*[T](a: var GMat4[T], i, j: int, v: T) =
+#   cast[var array[4*4, T]](a)[i * 4 + j] = v
+
+proc `[]=`*[T](a: var GMat2[T], i, j: int, v: T) =
+  case j:
+  of 0: a[i].x = v
+  of 1: a[i].y = v
+  else: discard
```
Comment (guzba) [inline src/vmath.nim:580]: except IndexError maybe

**Comment by treeform** [comment]
```nim c -r -d:release -d:vmathArrayBased .\tests\benchmark_svg.nim
svg decode ........................ 33.304 ms     42.224 ms    ±6.372   x117

nim c -r -d:release -d:vmathObjBased .\tests\benchmark_svg.nim
svg decode ........................ 27.000 ms     41.814 ms    ±5.096   x119

nim c -r -d:release .\tests\benchmark_svg.nim
svg decode ........................ 25.851 ms     36.101 ms    ±7.177   x139```

File: src/vmath.nim (hunk 2 above)
Comment (treeform) [reply src/vmath.nim]: remove

File: src/vmath.nim (hunk 3 above)
Comment (treeform) [reply src/vmath.nim]: doing pointe casting instead for speed

**Pull request merged**